<a href="https://colab.research.google.com/github/YaninaK/sales-forecast/blob/main/notebooks/04_Transformer_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sales forecast for 20 stores for 31 days' period.

## Transformer model

[EDA: omissions, clustering & Fourier transformations](
https://colab.research.google.com/drive/1udCgyEt7lusSY43lnduVDEnZvdvVEt9h?usp=sharing)

[Feature extraction](https://colab.research.google.com/drive/1H-vQoHnycp62YuZD1xrmIDjufpDu7ig3?usp=sharing)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!git clone -q https://github.com/YaninaK/sales-forecast.git
!pip install -r sales-forecast/requirements_Colab.txt -q 

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 358.2/358.2 KB 11.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 KB 12.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 136.2/136.2 KB 16.7 MB/s eta 0:00:00


In [3]:
%cd sales-forecast

/content/sales-forecast


In [4]:
import sys
import os

sys.path.append(os.getcwd())
sys.path.append(os.path.join(os.getcwd(), "src", "sales_forecast"))

In [5]:
import random
import pandas as pd
import numpy as np
import tensorflow as tf
import joblib

from data.make_dataset import load_data
from src.sales_forecast.models import train

from data.johnson_su_transformation import JohnsonSU
from data.train_test_datasets import get_train_dataset, get_test_dataset
from models.model_TF import get_model_TF

In [6]:
import warnings
warnings.filterwarnings('ignore')

In [7]:
PATH = '/content/drive/MyDrive/ML_projects/01_Time_series/sales_forecast/'
DEFAULT_RANDOM_SEED = 3

In [8]:
def set_all_seeds(seed=DEFAULT_RANDOM_SEED):
  random.seed(seed)
  os.environ['PYTHONHASHSEED'] = str(seed)
  np.random.seed(seed)
  tf.random.set_seed(seed)
  
set_all_seeds()

## 1. Data ingestion

In [9]:
data = load_data(PATH)

print(f'data.shape = {data.shape}\n')
data.head(2)

data.shape = (8230, 3)



id          dt  target
0   0  2016-01-02   747.8
1   0  2016-01-03   681.2

## 2. Data preprocessing pipeline

In [10]:
regenerate = True
if regenerate:
  X = train.data_preprocessing_pipeline(data)
else:
  X = np.load('data/04_feature/train_dataset.npy')

X.shape

Number of outliers = 8


(153, 20, 48)

## 3. Transformer model

### 3.1 Train dataset

In [11]:
X_train, y_train = get_train_dataset(X)
X_train.shape, y_train.shape

((900, 54, 48), (900, 27, 48))

### 3.2 Training of the model

In [12]:
model = get_model_TF(X_train.shape[1:])
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 54, 48)]     0           []                               
                                                                                                  
 layer_normalization (LayerNorm  (None, 54, 48)      96          ['input_1[0][0]']                
 alization)                                                                                       
                                                                                                  
 transformer_block (Transformer  (None, 54, 48)      62752       ['layer_normalization[0][0]']    
 Block)                                                                                           
                                                                                              

In [13]:
n_shops = 20
n_epochs = 50
batch_size = n_shops * 2
m = n_shops * 4

reduce_lr = tf.keras.callbacks.LearningRateScheduler(
    lambda epoch: 3e-2 * 0.95 ** epoch
)
model.compile(
    optimizer=tf.keras.optimizers.Adam(),
    loss=tf.keras.losses.Huber()
)
history = model.fit(
    X_train[:-m, :, :], y_train[:-m, :, :], 
    epochs=n_epochs,
    validation_data=(X_train[-m:, :, :], y_train[-m:, :, :]),
    batch_size=batch_size, 
    verbose=1,
    callbacks=[reduce_lr],
    shuffle=True,
    workers=-1,
    use_multiprocessing=True
)

Epoch 1/50
21/21 [==============================] - 27s 451ms/step - loss: 0.4094 - val_loss: 31.7494 - lr: 0.0300
Epoch 2/50
21/21 [==============================] - 7s 329ms/step - loss: 0.3511 - val_loss: 1.3434 - lr: 0.0285
Epoch 3/50
21/21 [==============================] - 9s 398ms/step - loss: 0.3403 - val_loss: 0.6471 - lr: 0.0271
Epoch 4/50
21/21 [==============================] - 7s 326ms/step - loss: 0.3294 - val_loss: 0.6117 - lr: 0.0257
Epoch 5/50
21/21 [==============================] - 8s 373ms/step - loss: 0.3187 - val_loss: 0.7136 - lr: 0.0244
Epoch 6/50
21/21 [==============================] - 7s 356ms/step - loss: 0.3081 - val_loss: 0.5099 - lr: 0.0232
Epoch 7/50
21/21 [==============================] - 7s 340ms/step - loss: 0.2975 - val_loss: 0.4762 - lr: 0.0221
Epoch 8/50
21/21 [==============================] - 7s 359ms/step - loss: 0.2884 - val_loss: 0.5979 - lr: 0.0210
Epoch 9/50
21/21 [==============================] - 7s 329ms/step - loss: 0.2798 - val_loss: 0

### 3.3 Inference

In [14]:
X_test, y_test = get_test_dataset(X)

In [15]:
pred = model.predict(X_test)
pred.shape

1/1 [==============================] - 3s 3s/step


(20, 27, 48)

### 3.4 Model evaluation

#### Mean absolute scaled error (MASE) 
https://en.wikipedia.org/wiki/Mean_absolute_scaled_error

The denominator of mean absolute scaled error is the mean absolute error of the one-step "seasonal naive forecast method" on the training set, which uses the actual value from the prior 30 days as the forecast: 
* F(t) = Y(t−30)

MASE is scale-free error metric. So, it can be applied to scaled data without extra iteration of inverse scaling.

In [16]:
pred.shape, y_test.shape

((20, 27, 48), (20, 27, 48))

In [17]:
forecast_error = abs(pred[:, :, 0] - y_test[:, :, 0]).mean()
mase = forecast_error / abs(y_test[:, :, 6]).mean()

print(f'Mean absolute scaled error = {round(mase.mean(), 4)}')

Mean absolute scaled error = 0.872


__Findings:__

* Out of 72 samples available for training (153 less input and output sequence length) 4 were used for validation of the model and 27 for testing. So, we had only 41 samples for training of the model and it is definately not sufficient.

* To get more accurate predictions, we need to retrain the model on the full set of data. It should improve the final results.

## 4. Final prediction

### 4.1. Dataset

In [18]:
X_train, y_train = get_train_dataset(X, final=True)
X_test = get_test_dataset(X, final=True)

X_train.shape, y_train.shape, X_test.shape

((1440, 54, 48), (1440, 27, 48), (20, 54, 48))

### 4.2. Retraining the model on full dataset

In [19]:
n_shops = 20
n_epochs = 60
batch_size = n_shops * 4
m = n_shops * 4

model = model = get_model_TF(X_train.shape[1:])
reduce_lr = tf.keras.callbacks.LearningRateScheduler(
    lambda epoch: 3e-2 * 0.95 ** epoch
)
model.compile(
    optimizer=tf.keras.optimizers.Adam(), 
    loss=tf.keras.losses.Huber()
)
history = model.fit(
    X_train[:-m, :, :], y_train[:-m, :, :], 
    epochs=n_epochs,
    validation_data=(X_train[-m:, :, :], y_train[-m:, :, :]),
    batch_size=batch_size, 
    verbose=1,
    callbacks=[reduce_lr],
    shuffle=True,
    workers=-1,
    use_multiprocessing=True,    
)

Epoch 1/60
17/17 [==============================] - 30s 775ms/step - loss: 0.3919 - val_loss: 237.0665 - lr: 0.0300
Epoch 2/60
17/17 [==============================] - 10s 574ms/step - loss: 0.3259 - val_loss: 5.3379 - lr: 0.0285
Epoch 3/60
17/17 [==============================] - 12s 661ms/step - loss: 0.3159 - val_loss: 0.8256 - lr: 0.0271
Epoch 4/60
17/17 [==============================] - 12s 689ms/step - loss: 0.3096 - val_loss: 0.7145 - lr: 0.0257
Epoch 5/60
17/17 [==============================] - 12s 691ms/step - loss: 0.3031 - val_loss: 0.5622 - lr: 0.0244
Epoch 6/60
17/17 [==============================] - 10s 618ms/step - loss: 0.2940 - val_loss: 0.3916 - lr: 0.0232
Epoch 7/60
17/17 [==============================] - 11s 598ms/step - loss: 0.2851 - val_loss: 0.3641 - lr: 0.0221
Epoch 8/60
17/17 [==============================] - 11s 676ms/step - loss: 0.2754 - val_loss: 0.3637 - lr: 0.0210
Epoch 9/60
17/17 [==============================] - 12s 685ms/step - loss: 0.2678 - va

### 4.3 Model evaluation on validation set

In [20]:
pred = model.predict(X_train[-m:, :, :])
pred[:, :, 0].shape,

3/3 [==============================] - 1s 63ms/step


((80, 27),)

In [21]:
forecast_error = abs(pred[:, :, 0] -  y_train[-m:, :, 0]).mean()
mase = forecast_error / abs( y_train[-m:, :, 6]).mean()

print(f'Mean absolute scaled error = {round(mase.mean(), 4)}')

Mean absolute scaled error = 0.4995


### 4.4. Inference

In [22]:
pred_final = model.predict(X_test)

1/1 [==============================] - 0s 74ms/step


### 4.5. Inverse transformation of predictions

In [23]:
scaler = joblib.load('data/03_primary/scaler_js.joblib')

In [24]:
arr_pred = pred_final[:, :, 0]
y_pred = pd.DataFrame(arr_pred).T
y_pred_rescaled = scaler.inv_transform(y_pred)
y_pred_rescaled.shape
     

(27, 20)

### 4.6. Generation of the final result

In [25]:
predict_example = pd.read_parquet(PATH + 'data/01_raw/sample_submit.parquet.gzip')
print(f'test.shape = {predict_example.shape}\n')
predict_example.head(2)

test.shape = (540, 3)



id          dt  target
0   0  2017-07-01     500
1   0  2017-07-02     500

In [26]:
n_shops = 20
for i in range(n_shops):
  predict_example.loc[predict_example['id'] == i, 'target'] = y_pred_rescaled.loc[:, i]

predict_example.to_parquet(
    'data/06_model_output/sales_forecast_tf.parquet.gzip', compression='gzip'
)
predict_example.head()

id          dt      target
0   0  2017-07-01  682.317587
1   0  2017-07-02  668.419487
2   0  2017-07-03  653.722555
3   0  2017-07-04  623.271131
4   0  2017-07-06  771.338321

## 5. Findings

Transformer model has not outperformed LSTM model. There is not enough data for transformers. The use of transformers could be justified on the longer time sequences.